### PR TITLE
Handle talking to older controllers

### PR DESCRIPTION
**Pull request by SimonRichardson**
Older controllers ignored channels and so should we when looking at
existing information.

The code just bails out if we're speaking to an older controller.

### CODE REVIEW

File: handlers.go [hunk 1]
```diff
@@ -247,6 +247,12 @@ func (r *resolver) allowCharmUpgrade(existingApp *Application, bundleApp *charm.
 	if existingApp.Revision == -1 {
 		return false, nil
 	}
+	// This handles the case that the existing application doesn't have a
+	// channel, so we're talking to an older controller. In that situation the
+	// older path way would ignore channel upgrades.
+	if existingApp.Channel == "" {
```
Comment (achilleasa) [inline handlers.go:253]: So we've got the model from an older controller. Why shouldn't we be allowed to do a channel upgrade if we want to given that it is the updated client that handles the changes anyway? Should we instead error out and allow a `--force` here?

Comment (SimonRichardson) [reply handlers.go:253]: Keeping consistency with older controllers. We can revisit this later if we want to?

Comment (achilleasa) [reply handlers.go:253]: But if I want to upgrade using a bundle with a specific charm channel, wouldn't an _older_ juju client let me go ahead anyway if I specify the revision (charmstore) + a channel?

Comment (SimonRichardson) [reply handlers.go:252]: Resolved

File: handlers.go [hunk 2]
```diff
@@ -247,6 +247,20 @@ func (r *resolver) allowCharmUpgrade(existingApp *Application, bundleApp *charm.
...
+	// channel, so we're talking to an older controller.
+	if existingApp.Channel == "" {
+		// No upgrade required.
+		if bundleApp.Channel == "" {
+			return false, nil
+		}
+		// If the bundle channel is not empty and we're not using force, then
+		// raise an error asking the user to supply force.
+		if !r.force && bundleApp.Channel != "" {
+			return false, errors.Errorf("upgrades not supported with unknown existing channels; use --force to override")
```
Comment (achilleasa) [inline handlers.go:260]:
```suggestion
			return false, errors.Errorf("upgrades not supported when the channel for the deployed application is unknown; use --force to override")
```

**Review by achilleasa — APPROVED** [review]
Left a minor wording improvement suggestion but other than that LGTM

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**